internal/mapping.go

```diff
@@ -164,3 +167,83 @@ func GetSpannerKeyName(srcKeyName string, schemaForeignKeys map[string]bool) str
 	schemaForeignKeys[spKeyName] = true
 	return spKeyName
 }
+
+// ResolveRefs resolves all table and column references in foreign key constraints
+// in the Spanner Schema. Note: Spanner requires that DDL references match
+// the case of the referenced object, but this is not so for many source databases.
+//
+// TODO: expand ResolveRefs to primary keys and indexes.
+func ResolveRefs(conv *Conv) {
+	for table, spTable := range conv.SpSchema {
+		spTable.Fks = resolveFks(conv, table, spTable.Fks)
+		conv.SpSchema[table] = spTable
+	}
+}
+
+// resolveFks returns resolved version of fks.
+// Foreign key constraints that can't be resolved are dropped.
+func resolveFks(conv *Conv, table string, fks []ddl.Foreignkey) (resolved []ddl.Foreignkey) {
+	for _, fk := range fks {
+		var err error
+		fk.Columns, err = resolveColRefs(conv, table, fk.Columns)
+		if err != nil {
+			conv.Unexpected(fmt.Sprintf("Can't resolve Columns in foreign key constraint: %s", err))
+			continue
+		}
+		fk.ReferTable, err = resolveTableRef(conv, fk.ReferTable)
+		if err != nil {
+			conv.Unexpected(fmt.Sprintf("Can't resolve ReferTable in foreign key constraint: %s", err))
+			continue
```

Comment:
continue is used at almost all the places where the Unexpected condition happens, and as we are anyway discarding the foreign key for any of the 3 reasons, it makes sense to just report one of the reason and continue.

Reply:
I would defer the decision here to @nevinheintze since I am not sure of the rationale behind not catching as many errors as possible in the report.

Reply:
First a bit of context on unexpected.

It is really intended as a mechanism for debugging the internals of harbourbridge itself, not for reporting errors to users. We typically use it when we encounter internal errors or unexpected conditions and instead of just generating a panic and exiting, we report it as "unexpected" and try to recover as best we can.

If we expect to hit these errors regularly, then we shouldn't use unexpected: instead we should figure out how to fix the problem and/or robustly report the issue to users.

In this particular case, we've encounter what looks like an invalid foreign key constraint. This should never be generated by mysqldump or pg_dump or from the infoschema tables. It could be generated by someone manually tweaking a dump file or writing one themselves. More generally, we don't focus in harbourbridge on helping users debug broken dump files. So I think unexpected is fine here.

And so generating the first unexpected and moving on is a reasonable tradeoff -- it is enough to help debug the issue from a harbourbridge perspective. Reporting all issues as unexpected wouldn't help very much, and it would add some complexity to this code --- we'd need to keep track of all of the errors and report them, and also keep track of whether we'd hit an error at all to know whether to keep this foreign key.

Reply:
So as Nevin mentioned  generating the first unexpected and moving on is a reasonable tradeoff , so there is no change required regarding this comment. marking this resolved.